airflow/utils/orm_event_handlers.py

```diff
@@ -23,18 +23,25 @@
 import traceback
 
 import sqlalchemy.orm.mapper
-from sqlalchemy import event, exc
+from sqlalchemy import __version__ as sqlalchemy_version, event, exc
 
 from airflow.configuration import conf
 
 log = logging.getLogger(__name__)
 
+SQL_ALCHEMY_V1 = sqlalchemy_version.startswith("1")
+
 
 def setup_event_handlers(engine):
     """Setups event handlers."""
     from airflow.models import import_all_models
 
-    event.listen(sqlalchemy.orm.mapper, "before_configured", import_all_models, once=True)
+    event.listen(
+        sqlalchemy.orm.mapper if SQL_ALCHEMY_V1 else sqlalchemy.orm.Mapper,
```

Comment:
Maybe it is better to create a compat module for SQLALchemy to handle this? Or maybe this sort of compat code can go into `airflow.utils.sqlalchemy` instead to avoid cluttering the code base.

Reply:
I'm not sure how to create such a module + it might be out of scope for this PR. I did move the check as suggested.

Reply:
This is easy - just wrap checking the sqlalchemy check in a function and move the fuction to a new `sqlalchemy.py` file in  `airflow.utils` folder - and I agree with @uranusjr - having it a separate util is a good idea. 

Se for example some checks in `airflow.utils.pydantic` module where we are doing very similar check - whether pydantic 1 or 2 or none is installed.

Reply:
@potiuk Do you not want to avoid code duplication? Why is it preferable to copy this entire block while changing one line over refactoring the version checking to a separate utility where the library and target version are inputs, then invoke from everywhere as needed?
```python
def is_pydantic_2_installed() -> bool:
    import sys

    from packaging.version import Version

    if sys.version_info >= (3, 9):
        from importlib.metadata import distribution
    else:
        from importlib_metadata import distribution
    try:
        return Version(distribution("pydantic").version) >= Version("2.0.0")
    except ImportError:
        return False
```

Reply:
Not sure what you are suggesting ?

Reply:
I did not suggest to copy the block (if that's what you think) - I was showing that we already have similar approach for Pydantic check (with more caveats because we also do fake imports there to make our code works without Pydantic).

Reply:
I suggest having a utility that's not specific to pydantic, pendulum sqlalchemy etc., something like:
```python
# import packaging, metadata, version, ...

def _get_lib_major_version(lib_name: str) -> int:
    ver: Version = parse_version(version(lib_name))
    return ver.major

def is_installed_lib_major_version(libname: str, version: int) -> bool:
    return _get_lib_major_version(lib_name) == version

# Then, in a TBD location:
def is_sqlalchemy_v1() -> bool:
    return is_installed_lib_major_version("sqlalchemy", 1)
```

Reply:
Feel absolutely free to make a refactor and propose a PR . This is how things work here. Someone sees an opportunity to improve things and proposes PRs :) , Maybe also find other places where such thing might be used (BTW. I think in this PR and in proposed solution you should merge the first tow mehtods, this will be way shorter and more readable.

Reply:
```python
def is_installed_lib_major_version(libname: str, version: int) -> bool:
    return parse_version(version(lib_name)).major == version
```

Or possibly even avoid  common another common utility but simply use that `one liner` everywhere (if it works also for various python versions - there is a good reason why we were checking for importlib in the previous solution)

Reply:
BTW, What's the difference between using distribution and version (and how it works for different Python versions?) did you check it?

Reply:
I did include the first of them. Can you suggest a suitable "general" place for them? Should I add another file to the utils folder?

Reply:
Ah yeah - seems we can use importlib.metadata version directly - the importlib_metadata imports are not needed as we are not using an features that would not be available in Python 3.8  - so yes, feel free to simplify all that (probably in a separate PR).

Reply:
I really want to do something like this, but don't know where to put it. The main reason for not wanting to append it to an existing file is because I want it to be as standalone as possible, so that it cannot ever cause circular imports.

```python
# e.g. airflow/utils/dependency_versions.py
from importlib.metadata import version as get_version_string

from packaging.version import parse as parse_version

def is_installed_lib_major_version(lib_name: str, major_version: int) -> bool:
    """Check if an installed library has the specified major version."""   
    return parse_version(get_version_string(lib_name)).major == major_version
```
Then wherever this is needed will either have a helper wrapper or be used directly
```python
# airflow/utils/sqlalchemy.py, or airflow/settings.py
from airflow.utils.dependency_versions import is_installed_lib_major_version

# OPTION 1:
def is_sqlalchemy_v1() -> bool:
    return is_installed_lib_major_version("sqlalchemy", 1)

...
if is_sqlalchemy_v1():
    # do something

# OPTION 2:
if is_installed_lib_major_version("sqlalchemy", 1):
    # do something
```

----

BTW, why is `airflow.utils.orm_event_handlers` a separate file and not part of `airflow.utils.sqlalchemy`?

Reply:
create a separate file -propose a name, or find a good one. Generally we expect to autonomously propose somethign by looking at the code and doing things similarly. If it looks good, we will accept it, if not, we will suggest changes.

> BTW, why is airflow.utils.orm_event_handlers a separate file and not part of airflow.utils.sqlalchemy?

Because it's author chose it to be in a sparate package I guess?  There are no strict rules for that  and is (as many naming decisions) subject to individual person and reviewer. But if you have an idea and need to introduce some rules for naming that will be easy to answer question `why` then  I think you should propose it.

Gennerally - it's not a science , it's art and that art depends very much with artist you interact with. You have to accept ambiguities, some level of uncertainty and show initiative on your own (and accept the fact that commiters might have different (sometimes between them) opinions, it's also up to you to figure out how to get to consensus for names etc.

Reply:
`Cache invalidation, naming things and off-by one mistakes are the two most difficult things in comuputer science`

Reply:
Thank you for the insights @potiuk.

I honestly don't know what to do about this PR anymore, and it is frustrating. It was supposed to be a very minor change that got blown out of proportion. The PR seems to be "good enough" according to the reviewers, and yet it's stuck.

What do I need to do to get it merged? Keep rebasing it and hope for a slot where nothing else merges in the meantime? Resolve all threads unilaterally? Change the code?